traits/ctraits.c

```diff
@@ -310,9 +310,27 @@ fatal_trait_error ( void ) {
 +----------------------------------------------------------------------------*/
 
 static int
-invalid_attribute_error ( void ) {
+invalid_attribute_error ( PyObject * name ) {
 
-    PyErr_SetString( PyExc_TypeError, "attribute name must be string" );
+    // Build error message by collecting the type of the object received
+    PyTypeObject* obj_type = name->ob_type;
+    const char* type_name = obj_type->tp_name;
+    PyObject * ob_repr = PyObject_Repr(name);
```

Comment:
You've created a new reference here;  there needs to be a corresponding `Py_DECREF` at some point before the end of the function to compensate.